crypto/store/loader_file.c

```diff
@@ -1128,6 +1176,64 @@ static int file_name_to_uri(OSSL_STORE_LOADER_CTX *ctx, const char *name,
     return 1;
 }
 
+static int file_name_check(OSSL_STORE_LOADER_CTX *ctx, const char *name)
+{
+    const char *p = NULL;
+
+    /* If there is no search criteria, all names are accepted */
+    if (ctx->_.dir.search_name[0] == '\0')
+        return 1;
+
+    /* If the expected type isn't supported, no name is accepted */
+    if (ctx->expected_type != 0
+        && ctx->expected_type != OSSL_STORE_INFO_CERT
+        && ctx->expected_type != OSSL_STORE_INFO_CRL)
+        return 0;
+
+    /*
+     * First, check the basename
+     */
+    if (strncasecmp(name, ctx->_.dir.search_name,
+                    sizeof(ctx->_.dir.search_name) - 1) != 0
+        || name[sizeof(ctx->_.dir.search_name) - 1] != '.')
+        return 0;
+    p = &name[sizeof(ctx->_.dir.search_name)];
+
+    /*
+     * Then, if the expected type is a CRL, check that the extension starts
+     * with 'r'
+     */
+    if ((ctx->expected_type == 0
+         || ctx->expected_type == OSSL_STORE_INFO_CRL)
+        && *p++ != 'r')
+        return 0;
```

Comment:
Doesn't `expected_type == 0` mean allow any type? So why are we failing if it is set to zero here and type isn't 'r'?

Reply:
Don't we need a similar check for `OSSL_STORE_INFO_CERT`?

Reply:
Ah, yes, the logic is flawed.  So basically, there are three cases:

1.  expected type == 0, 'r' may or may not be present and that's ok.
2.  expected type == OSSL_STORE_INFO_CRL, 'r' *must* be present.
3.  expected type is anything else, 'r' *must not* be present.

Reply:
Fixed, I hope.